background/objectoriented.tex

```diff
@@ -1,210 +1,333 @@
-      self->data = 0;
-    }
-
-    uint32_t Sensor_Measure(Sensor *self) {
-      return self.data;
-    }
-
-    int main() {
-
-      Sensor s1;
-      Sensor_New(&s1);
-      printf("%d\n", Sensor_Measure(&s1));
-
-      Sensor* s2 = (Sensor*)malloc(sizeof(Sensor));
-      Sensor_New(s2);
-      printf("%d\n", Sensor_Measure(s2));
-      free(s2);
-
-      return 0;
-    }
-  \end{minted}
-  \caption{Sensor defined in C}
-  \label{lst:back:oo:c}
-\end{listing}
-
-C does not provide any syntax for dealing with method calls on objects, but the object oriented paradigm can be used through convention.
-\autoref{lst:back:oo:c} shows a definiton for a Sensor class with constructor function and member function in the C programming language.
-Here we see the method call mechanism explicit in the main function where the a reference to the object has to be passed explicit as the first argument.
-Memory allocation is also handled explicit through malloc when allocating a object on the heap.
-The pattern can be changed to expose a New function that returns a reference to a heap allocated object, but this either prevents stack allocated objects or leads to alot of to to handle it.
-The main function also contains a explicit free of the heap allocated memory which is not present in \autoref{lst:back:oo:rust}.
-This is due to the ownership rules of Rust explained in \autoref{sec:back:rust:own}.
-
-\subsection{Object Oriented Hardware components}
-
-In the previous section we saw how the OO-paradigm is implemented in Java, Rust and C.
-We can see that in essence the paradigm consists of a datatype with a number of fields, a collection of function that operates on the fields given a reference to a object of this datatype.
-This paradigm can be directly applied to software components in a embedded system such as a circular buffer.
-
-Hardware components in a embedded system as the Gecko Mircocontroller are often exposed as memory mapped devices.
-This implies for instance that the control registers for a devices such as an Analog-To-Digital is located in the address space of the CPU and can be read from and written to with the same instruction and language constructs as the rest of the memory.
-\autoref{fig:back:memorymapped} shows the control registers for the device which are placed continuously in memory.
-
+In its essense Object Oriented programming focues on organizing a computer program by looking at the data the program operates on, grouping related data into objects and defining methods that operate on the data contained within the objects.
+The paradigms essensial concept be applied to embedded C programming even though the language does not define language features to aid the design.
+This section looks at how Object Oriented programming is used in conjunction with memory mapped I/O in embedded programming.
+We look at the memory layout of a memory mapped Analoge-To-Digital and see how this can be represented as an object.
```

Comment:
Analogue-To-Digital Converter?

Evt legge til `\gls{adc}` i `abbreviations.tex`